text/0000-decouple-dependencies.md

```diff
@@ -0,0 +1,403 @@
+matching is generally simpler that way, however, if a dependency does not
+follow semver you may use regular expressions to match the versions that are
+compatible with the buildpack.
+
+If any semver range or regular expression matches then it can be assumed that
+the buildpack is compatible. If no range matches then it can be assumed that
+the version is not compatible.
+
+For example:
+```toml
+[[metadata.validations]]
+dependency-id = "com.example.jre"
+supported = [ "8.0.*", "11.0.*", "17.0.*" ]
+
+[[metadata.validations]]
+dependency-id = "com.example.nodejs"
+supported = [ "^16.0", "^17.0", "^18.0" ]
+
+[[metadata.validations]]
+dependency-id = "com.example.tomcat"
+supported = [ "8\.5\.\d+", "9\.0\.\d+", "10\.0\.\d+" ]
+type = “regex”
+```
+
+A buildpack is encouraged to be as permissive as possible. This ensures that a
+buildpack author won’t have to frequently update this metadata. This should be
+balanced by buildpack authors to provide compatibility guarantees with the
+tools required to run the buildpack and for the software it is installing to
+run. Generally, we believe that most buildpacks will be compatible with the
+major versions of software they presently support and that new major versions
+of dependencies should be tested and validations should be expanded after
+testing.
+
+## Metadata Distribution
+There is no prescribed method for distributing metadata. It could be done in a
+variety of ways, including HTTPS/SFTP distributed archives, `rsync` of remote
+directories, or even distributed as an image through an image registry.
+
+For the Paketo project, this proposal suggests distributing metadata through
+images in an image registry. This allows the project to use the existing image
+registry to distribute the metadata. Image registries also have inherent
+properties that help with security, like an image cannot be modified without
+creating a new hash for the image and images can be signed (signing is out of
+scope for this proposal). In addition, images are easily versioned such that
+users can hold back updates to dependencies if desired and are easily cached.
+
+The contents of the image will contain the directory structure defined in
+[Metadata Format](#metadata-format). There should not be a top-level directory
+added, so the root of the image should contain all of the directories created
+as top-level organization names, like `com` or `org`. All of the metadata is to
+be included in a single layer. Updates to metadata will require downloading the
+entire layer again, however, it is a single layer and the size is expected to
+be small so this should be very fast.
+
+Further, this proposal suggests subdividing metadata images by project
+sub-team. Each sub-team will be given a unique reverse domain name like
+`io.paketo.java` and `io.paketo.utilities`. In this way, the project’s metadata
+can be easily combined without having conflicts.
+
+This allows users to pick and choose the metadata that’s relevant to their
```

Comment:
Who are "users" in this context? Who is able to pick and choose metadata? I assume this refers to Buildpack authors?

Reply:
It could be from multiple perspectives.

- A buildpack author consuming metadata and dependencies would not really care. They get whatever is provided to them. Unless you're talking a composite buildpack in which case the buildpack author could have another buildpack in the order group that provides metadata. So a language team could control dependencies this way.

- A builder author could provide metadata. If you were packaging up a custom builder you could override the dependency metadata with the selection you want available to your builders users.

- an end user could volume mount in dependency metadata to customize what is usable at build time.

I suspect Paketo language family teams will probably use a buildpack to add metadata to the build container. They would want to skip installing metadata if it exists already like if it's on the builder or volume mounted into the container. That gives builder authors or end users the ability to override things with a custom set of metadata.

Reply:
Yes, that makes sense. Maybe it's worth adding that context to the RFC - the expansion of "users" to those use-cases was very helpful for me.

Reply:
agreed, this explanation was helpful and it would be great to have in the text

Reply:
OK, I'll add it in.